content/blog/using-docker-wrong-my-journey-to-a-better-container.md

```diff
@@ -0,0 +1,53 @@
+
+One of the primary benefits of Docker should be that, if I create a Docker container, I will be able run it on all my systems, and if I hand that container to someone else, it should run the same way, thereby being a deterministic system. Now, let me say there are always caveats, but this should be a large goal for Docker users.
+
+Where did I go wrong? Well, my main issue was using my old code from before the MapR PACC.  In this code, I took a number of directories for StreamSets, including logs, conf (config), and others, and placed it on the local system.  When I ran this on a MapR node in my cluster, it worked great, because all the users that I would use inside the container were also on my MapR nodes running Docker. Thus, I could set up permissions on those local directories, and when the container ran, the user inside the container had proper filesystem permissions to the directories required by StreamSets.
+
+![](/img/docker-host.png)
+
+However, when I shared my repository with a colleague at MapR, they reported it didn’t work on their system. I was baffled: it should work; it’s Docker! That’s the whole point of using containers. Well, my colleague was not running the container on MapR nodes; instead, it was running on a client machine. This client machine did NOT have the same users as the MapR cluster. Thus there were a number of issues that my scripts ran into with creating the configuration directories, settings permissions, and even running StreamSets.  I made assumptions in my original Docker, and by doing so, made a container that ran differently depending on the execution system.
+
+This was an eye-opener for me.  From one perspective, I had tried to get StreamSets running in the PACC by using old knowledge and approaches in an effort to expedite the process. By doing so, I hampered the usability, portability, and shareability of my container. To that end, I’ve created another GitHub repository:
+
+<https://github.com/johnomernik/maprssss>
+
+Here is my first attempt to “get it right.”  Instead of volume mounting the configuration directories for StreamSets to the local filesystem, I use the MapR POSIX Client built into the PACC to put the configuration directly on MapR XD.  This ensures that the user in the container always has access and that the config files are available to StreamSets anywhere they run.  
+
+![](/img/docker-host-2.png)
+
+This allowed me to simplify my startup scripts and ensure a much more clean runtime environment, independent of the system I was running the container on. This is a much better approach and really shows the power of using MapR with Docker containers.  I am still learning more about the PACC, but as I learn, I wanted to take the time to share my follies, so others can learn as well.  If you take the time to review my repositories here and find ways to improve them even more, I am all ears! Leave a comment, or post a GitHub issue.
```

Comment:
Suggestion to make "here" a hotlink to the GitHub repository so it reads:
If you take the time to review my repositories [here](https://github.com/johnomernik/maprssss) and find ways to improve even more...

Also, it is "my repositories here" or "my repository here" ? (singular or plural)?